pvsystemprofiler/latitude_study.py

```diff
@@ -0,0 +1,111 @@
+
+        self.data_handler = data_handler
+        self.phi_true_value = lat_true_value
+        self.phi_true_value = lat_true_value
+        if not data_handler._ran_pipeline:
+            print('Running DataHandler preprocessing pipeline with defaults')
+            self.data_handler.run_pipeline()
+        self.data_matrix = self.data_handler.filled_data_matrix
+        self.raw_data_matrix = self.data_handler.raw_data_matrix
+        self.day_of_year = self.data_handler.day_index.dayofyear
+        self.num_days = self.data_handler.num_days
+        self.daily_meas = self.data_handler.filled_data_matrix.shape[0]
+        self.data_sampling = self.data_handler.data_sampling
+        self.boolean_daytime = None
+        self.discrete_latitude = None
+        self.hours_daylight = None
+        self.delta = None
+        self.residual = None
+        # Results
+        self.results = None
+
+    def run(self, threshold_method=('raw data matrix', 'filled data matrix'),
+            daylight_method=('raw daylight', 'sunrise-sunset'),
+            threshold=None, verbose=True):
+        threshold_method = np.atleast_1d(threshold_method)
+        daylight_method = np.atleast_1d(daylight_method)
+
+        thres = threshold
+        if threshold is None:
+            thres = 0.001*np.ones(len(threshold_method)*len(daylight_method))
+        self.make_delta()
+
+        results = pd.DataFrame(columns=['latitude', ' threshold', 'threshold matrix'])
+        counter = 0
+        for matrix_ix, matrix_id in enumerate(threshold_method):
+            for daylight_method_id in daylight_method:
+                dtt = thres[counter]
+                dlm = daylight_method_id
+                met = threshold_method[matrix_ix]
+                lat_est = self.estimate_latitude(matrix_id, daytime_threshold=dtt, daylight_method=dlm)
+
+                results.loc[counter] = [lat_est, dtt, met]
+                counter += 1
+        if self.phi_true_value is not None:
+            results['residual'] = self.phi_true_value - results['latitude']
+
+        self.results = results
+
+    def estimate_latitude(self, matrix_id=None, daytime_threshold=0.001, daylight_method=('sunrise-sunset',
+                                                                                          'raw daylight')):
+        if matrix_id in ('raw data matrix', 'raw_data_matrix', 'raw'):
+            data_in = self.raw_data_matrix
+        if matrix_id in ('filled data matrix', 'filled_data_matrix', 'filled'):
+            data_in = self.data_matrix
+
+        if daylight_method in ('sunrise-sunset', 'sunrise sunset'):
+            self.hours_daylight = self.calculate_hours_daylight(data_in, daytime_threshold)
+        if daylight_method in ('raw_daylight', 'raw daylight'):
+            self.hours_daylight = self.calculate_hours_daylight_raw(data_in, daytime_threshold)
+
+        self.discrete_latitude = np.degrees(np.arctan(- np.cos(np.radians(15 / 2 * self.hours_daylight)) /
```

Comment:
This is the core of the "latitude estimation algorithm," the equation that relates daylight hours to the latitude. This should be moved into it's own module here: `pvsystemprofiler/alorithms/latitude/direct_calculation.py`. You can look at `pvsystemprofiler/alorithms/longitude/direct_calculation.py` as an example.